### PR TITLE
Spiders for portal da transparência

### LINKED ISSUE #353: Create base spider for portaldatransparencia.com.br
Several cities use portaldatransparencia.com.br to host their gazettes, e.g [BA/Varzea do Poço](http://ba.portaldatransparencia.com.br/prefeitura/varzeadopoco/).

It would be nice to create a base spider like FecamBaseSpider, so that it would be easy to create spiders for these cities by just inheriting the base spider, setting territory ID, url for the city and so on.

A list of all cities that use this platform can be found here: http://atmtec.org.br/#publicacao

Note that some cities only have old gazettes there, but some are still using it. If one  decides to also create a few spiders in the PR related to this issue, it's worth  checking which cities are actively using the platform.

### PR DESCRIPTION
**Pull request by rodbv**
Closes #353 

Creates a base spider for Portal da Transparência and adds spiders for the cities that are currently using this platform

### CODE REVIEW

File: data_collection/gazette/spiders/base/portal_da_transparencia.py
Comment (rennerocha) [inline data_collection/gazette/spiders/base/portal_da_transparencia.py:17]: **suggestion**: I don't see any gain on  creating this constant just to use it in the next line. If the value is not used in a different location, it just add some context change when we are reading the code. Also, if the webpage changes its content, it is very unlikely that we will just need to update this specific selector to fix it (as probably the entire page changed).

```
editions = response.css("div.box-edicao")
for edition in editions:
    ....
```
or simply:
```
for issue in response.css("div.box-edicao"):
    ...
```

File: data_collection/gazette/spiders/base/portal_da_transparencia.py
Comment (rennerocha) [inline data_collection/gazette/spiders/base/portal_da_transparencia.py:13]: **question**: How did you find this date?

File: data_collection/gazette/spiders/base/portal_da_transparencia.py
Comment (rennerocha) [inline data_collection/gazette/spiders/base/portal_da_transparencia.py:39]: **suggestion**: We know that this selector is getting the date, as it is used in the following line. I don't think this increases the legibility of the code.

File: data_collection/gazette/spiders/base/portal_da_transparencia.py
Comment (rennerocha) [inline data_collection/gazette/spiders/base/portal_da_transparencia.py:43]: **suggestion**: This method is too short and can be replaced by one line in `parse` method. Doing that, we remove the need to scroll the code when reading it later.

```
gazette_file_url = response.urljoin(issue.css("a.btn_visualizar::attr(href)").get())
```

File: data_collection/gazette/spiders/base/portal_da_transparencia.py
Comment (rennerocha) [inline data_collection/gazette/spiders/base/portal_da_transparencia.py:50]: **suggestion**: You can use a regex directly:

This will return the string of the edition number:
```
edition_number = issue.css("span.edicao > strong ::text").re_first("Edição (\d+)")
```

**Review by rennerocha — CHANGES_REQUESTED** [review]
Added a few comments. Please let me know if you have any question.

File: data_collection/gazette/spiders/base/portal_da_transparencia.py
Comment (rodbv) [reply data_collection/gazette/spiders/base/portal_da_transparencia.py:17]: I like to declare these constants so they inform the reader what it's meant for (in this case, get the edition data). Becomes a bit less of a "magic string"

Comment (rennerocha) [reply data_collection/gazette/spiders/base/portal_da_transparencia.py:17]: So `editions = response.css("div.box-edicao")` makes more sense in my opinion. It clearly indicates what is the purpose of the selector content. What bothers me most on this approach is that I need to change the context (move my focus to a different line) to see what exactly I am searching. Considering that we expect to have selector strings in a spider code, they are not magic strings, they are just part of the process to get the information. AS you are not reusing the string in other places, it makes no much sense to have an extra variable just for that.

Comment (rodbv) [reply data_collection/gazette/spiders/base/portal_da_transparencia.py:17]: Thanks 👍

File: data_collection/gazette/spiders/base/portal_da_transparencia.py
Comment (rodbv) [reply data_collection/gazette/spiders/base/portal_da_transparencia.py:13]: It's just a date that is "safe enough" in the context of this platform as there are no gazettes prior to that, but on the other hand it has no performance-related downsides on the way they retrieve the gazettes.

File: data_collection/gazette/spiders/base/portal_da_transparencia.py
Comment (rodbv) [reply data_collection/gazette/spiders/base/portal_da_transparencia.py:39]: Done

File: data_collection/gazette/spiders/base/portal_da_transparencia.py
Comment (rodbv) [reply data_collection/gazette/spiders/base/portal_da_transparencia.py:50]: Awesome thanks

File: data_collection/gazette/spiders/base/portal_da_transparencia.py
Comment (rodbv) [reply data_collection/gazette/spiders/base/portal_da_transparencia.py]: Good point thanks

**Comment by rodbv** [comment]
Hey there, is there anything I need to do to get this PR closed? I've implemented the suggested changes and I'm not sure if I have to flag it somewhere.

File: data_collection/gazette/spiders/ap_tartarugalzinho.py
Comment (anapaulagomes) [inline data_collection/gazette/spiders/ap_tartarugalzinho.py:7]: The last edition dates from 1/Janeiro/2021. Could you take a look at the current gazettes' website?

File: data_collection/gazette/spiders/ba_botupora.py
Comment (anapaulagomes) [inline data_collection/gazette/spiders/ba_botupora.py:7]: The last edition dates from 22/Março/2021. Could you take a look at the current gazettes' website?

File: data_collection/gazette/spiders/ba_conde.py
Comment (anapaulagomes) [inline data_collection/gazette/spiders/ba_conde.py:7]: The last edition dates from 30/Dezembro/2020. Could you take a look at the current gazettes' website?

File: data_collection/gazette/spiders/ba_ibitiara.py
Comment (anapaulagomes) [inline data_collection/gazette/spiders/ba_ibitiara.py:7]: The last edition dates from 20/Janeiro/2012. Could you take a look at the current gazettes' website?

File: data_collection/gazette/spiders/ba_santa_rita_de_cassia.py
Comment (anapaulagomes) [inline data_collection/gazette/spiders/ba_santa_rita_de_cassia.py:8]: The last edition dates from 31/Dezembro/2020. Could you take a look at the current gazettes' website?

**Review by anapaulagomes — CHANGES_REQUESTED** [review]
@rodbv thanks for your contribution! Could you please update your branch and also add the cities to our CITIES.md? I promise I'll be back to you as soon as you update it. I'm still running your spiders but so far so good. I'll comment here if something pops up.

I left a few comments about the spiders. I think that would be nice to at least creating new issues for the portals that are outdated and will need new spiders.

A question for the maintainer's team: how are we handling different websites for the same city? @giuliocc @jvanz (not sure whom else mention here)

File: data_collection/gazette/spiders/ap_tartarugalzinho.py
Comment (rodbv) [reply data_collection/gazette/spiders/ap_tartarugalzinho.py:7]: They've moved to this URL http://tartarugalzinho.ap.gov.br/diario_oficial, which also seems pretty straightforward to scrap. I wonder what's the procedure for when a city needs 2 or more spiders depending on date. Just add some "if" statements to point to one or another scraping function?

File: data_collection/gazette/spiders/ba_angical.py
Comment (rodbv) [reply data_collection/gazette/spiders/ba_angical.py:7]: They've moved to this URL http://www.acessoinformacao.com.br/ba/angical/diario
Not sure if there's a generic spider for this site already, if not it could be another opportunity to add several cities at one.

File: data_collection/gazette/spiders/ba_conde.py
Comment (rodbv) [reply data_collection/gazette/spiders/ba_conde.py:7]: It's here now: http://botupora.ba.gov.br/diario_oficial
Underneath, they seem to be a platform called http://procedebahia.com.br (if there are other spiders on top of this platform, it may be possible to reuse code)

File: data_collection/gazette/spiders/ba_ibitiara.py
Comment (rodbv) [reply data_collection/gazette/spiders/ba_ibitiara.py:7]: https://www.ibitiara.ba.gov.br/site/diariooficial

File: data_collection/gazette/spiders/ba_paramirim.py
Comment (rodbv) [reply data_collection/gazette/spiders/ba_paramirim.py:7]: it's here http://procedebahia.com.br/ba/paramirim (same platform as Botupora)

File: data_collection/gazette/spiders/ba_santa_maria_da_vitoria.py
Comment (rodbv) [reply data_collection/gazette/spiders/ba_santa_maria_da_vitoria.py:8]: They've moved to http://pmsantamariadavitoriaba.imprensaoficial.org/ultimos-diarios/

File: data_collection/gazette/spiders/ba_santa_rita_de_cassia.py
Comment (rodbv) [reply data_collection/gazette/spiders/ba_santa_rita_de_cassia.py:8]: They seem to still be using the platform, but with the URL https://doem.org.br/ba/santaritadecassia

File: data_collection/gazette/spiders/ba_varzea_da_roca.py
Comment (rodbv) [reply data_collection/gazette/spiders/ba_varzea_da_roca.py:7]: It's now at http://impublicacoes.org/agenda/index.php?month=1&year=2021&type=pref&municipio=TWprek16QTFPUT09

File: data_collection/gazette/spiders/ap_tartarugalzinho.py
Comment (anapaulagomes) [reply data_collection/gazette/spiders/ap_tartarugalzinho.py:7]: I think so (at least it is what is done in other spiders). But feel free to open an issue, so this can be tackled later.

File: data_collection/gazette/spiders/ba_rio_do_pires.py
Comment (anapaulagomes) [inline data_collection/gazette/spiders/ba_rio_do_pires.py:6]: About the execution:

Total of gazettes: 1746
First date: 2004-01-16 
Last date: 2021-08-13 ✅ 
Average of gazettes per day (6419 days): 0.3
Average of gazettes by month (211 months): 8.3
Average of gazettes by year (17 years): 102.7

Found 30 errors in log:
`sqlite3.IntegrityError: UNIQUE constraint failed: gazettes.territory_id, gazettes.date, gazettes.file_checksum`

File: data_collection/gazette/spiders/ba_santa_maria_da_vitoria.py
Comment (anapaulagomes) [reply data_collection/gazette/spiders/ba_santa_maria_da_vitoria.py:8]: About the execution:

Total of gazettes: 3520
First date: 2003-08-28 
Last date: 2021-06-14
Average of gazettes per day (6500 days): 0.5
Average of gazettes by month (214 months): 16.4
Average of gazettes by year (18 years): 195.6

Found 152 errors in log:
`sqlite3.IntegrityError: UNIQUE constraint failed: gazettes.territory_id, gazettes.date, gazettes.file_checksum`

**Review by anapaulagomes — COMMENTED** [review]
Still running the spiders 😭  hahaha but I left a few comments on the results. So far so good. I just had to update the branch. In the future, we may have to add `end_date` but I'm going to leave this to another PR.

The errors from the logs are related to unique constraints. I don't think it is a problem, since the numbers look good.

**Review by anapaulagomes — APPROVED** [review]
I finally finished the tests. 😅  I documented the errors and it seems something recurrent. I don't see problems here but you let me know.

It looks good to me! After you update the branch, we are ready to go! 🎉 
As a follow-up task, would be nice to create issues for the cities that changed their websites.

**Comment by rennerocha** [comment]
Closed as stale.

**Pull request closed without merging**
